### PR TITLE
Missing data handling for mixedlm

### LINKED ISSUE #3248: statsmodels.formula.api.mixedlm does to know how to handle missing data
Hi there,
The formula interface for MixedLM does not know how to handle missing values. It also does not handle categorical data very well, too.

Minimal Example

```
import numpy as np
import pandas as pd
from pandas import Series, DataFrame
from patsy import dmatrix
from patsy import dmatrices
import statsmodels.api as sm
import statsmodels.formula.api as smf

x = Series(random.randint(0,2,50), dtype='category')
x.cat.categories = ['foo', 'bar']

y = Series(random.randint(0,3,50), dtype='category')
y.cat.categories = ['low', 'medium', 'high']

z = Series(random.randint(0,2,50), dtype='category')
z.cat.categories = ['male', 'female']

data = DataFrame({
    'response': random.rand(50),
    'assistant': x,
    'risk' : y,
    'sex': z,
    'age': random.randint(20,60,50)})
```

Which gives us:

```
In [59]: data.head()
Out[59]: 
   age assistant  response    risk     sex
0   38       foo  0.524707  medium  female
1   38       foo  0.902656    high  female
2   39       foo  0.934783     low    male
3   53       foo  0.136764    high    male
4   34       foo  0.013332     low    male
```

The grouping in the data frame is a categorical value, which mixedlm complains about.

```
In [65]: fit = smf.mixedlm('response ~ sex + age + risk', groups=data['assistant'], data=data).fit()
/home/moebius/software/python-environments/for/lib/python3.5/site-packages/statsmodels/regression/mixed_linear_model.py:160: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  self._params = np.zeros(self.k_tot)
/home/moebius/software/python-environments/for/lib/python3.5/site-packages/statsmodels/regression/mixed_linear_model.py:1717: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)
```

If we set some covariates to missing, we get an error.

```
In [66]: data.risk[10] = np.nan

In [67]: fit = smf.mixedlm('response ~ sex + age + risk', groups=data['assistant'], data=data).fit()
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-67-decde0c23284> in <module>()
----> 1 fit = smf.mixedlm('response ~ sex + age + risk', groups=data['assistant'], data=data).fit()

/home/moebius/software/python-environments/for/lib/python3.5/site-packages/statsmodels/regression/mixed_linear_model.py in from_formula(cls, formula, data, re_formula, subset, *args, **kwargs)
    651                                                subset=None,
    652                                                exog_re=exog_re,
--> 653                                                *args, **kwargs)
    654 
    655         # expand re names to account for pairs of RE

/home/moebius/software/python-environments/for/lib/python3.5/site-packages/statsmodels/base/model.py in from_formula(cls, formula, data, subset, *args, **kwargs)
    148         kwargs.update({'missing_idx': missing_idx,
    149                        'missing': missing})
--> 150         mod = cls(endog, exog, *args, **kwargs)
    151         mod.formula = formula
    152 

/home/moebius/software/python-environments/for/lib/python3.5/site-packages/statsmodels/regression/mixed_linear_model.py in __init__(self, endog, exog, groups, exog_re, use_sqrt, missing, **kwargs)
    537 
    538         # Split the data by groups
--> 539         self.endog_li = self.group_list(self.endog)
    540         self.exog_li = self.group_list(self.exog)
    541         self.exog_re_li = self.group_list(self.exog_re)

/home/moebius/software/python-environments/for/lib/python3.5/site-packages/statsmodels/regression/mixed_linear_model.py in group_list(self, array)
    671         if array.ndim == 1:
    672             return [np.array(array[self.row_indices[k]])
--> 673                     for k in self.group_labels]
    674         else:
    675             return [np.array(array[self.row_indices[k], :])

/home/moebius/software/python-environments/for/lib/python3.5/site-packages/statsmodels/regression/mixed_linear_model.py in <listcomp>(.0)
    671         if array.ndim == 1:
    672             return [np.array(array[self.row_indices[k]])
--> 673                     for k in self.group_labels]
    674         else:
    675             return [np.array(array[self.row_indices[k], :])

IndexError: index 49 is out of bounds for axis 1 with size 49
```

The problem is obviously that `group=data['assistant']` is now too long and does not fit the design matrix build by patsy any more.

Q: How does the formula api make sure that the order in data['assistant'] and the design matrix produced by patsy always match? I don't know what patsy does under the hood, but is it allowed to change the order in which samples are given to it?

If there are missing values in data['assistant'] it also breaks.

```
In [69]: data.dropna(inplace=True)
In [70]: data.assistant[21] = np.nan
In [71]: fit = smf.mixedlm('response ~ sex + age + risk', groups=data['assistant'], data=data).fit()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-71-decde0c23284> in <module>()
----> 1 fit = smf.mixedlm('response ~ sex + age + risk', groups=data['assistant'], data=data).fit()

/home/moebius/software/python-environments/for/lib/python3.5/site-packages/statsmodels/regression/mixed_linear_model.py in from_formula(cls, formula, data, re_formula, subset, *args, **kwargs)
    651                                                subset=None,
    652                                                exog_re=exog_re,
--> 653                                                *args, **kwargs)
    654 
    655         # expand re names to account for pairs of RE

/home/moebius/software/python-environments/for/lib/python3.5/site-packages/statsmodels/base/model.py in from_formula(cls, formula, data, subset, *args, **kwargs)
    148         kwargs.update({'missing_idx': missing_idx,
    149                        'missing': missing})
--> 150         mod = cls(endog, exog, *args, **kwargs)
    151         mod.formula = formula
    152 

/home/moebius/software/python-environments/for/lib/python3.5/site-packages/statsmodels/regression/mixed_linear_model.py in __init__(self, endog, exog, groups, exog_re, use_sqrt, missing, **kwargs)
    528         # list of arrays, corresponding to the groups.
    529         group_labels = list(set(groups))
--> 530         group_labels.sort()
    531         row_indices = dict((s, []) for s in group_labels)
    532         for i,g in enumerate(groups):

TypeError: unorderable types: str() < float()
```

A work around is to subset the data frame to only contain the variables needed in the model fit and drop all missings. It is just not very practical during model building and selection when a lot of variables are added and removed on the fly.

Maybe it would be more intuitive to define the grouping as a formula, too, say `groups='~assistant'` and that mixedlm works out the details by itself? I don't know. You are the experts.

Thank you for your help, your efforts, and a great package!

### PR DESCRIPTION
**Pull request by kshedden**
Closes #3248

### CODE REVIEW

**Comment by kshedden** [comment]
Actually may not fully close #3248, I didn't look into handling Pandas categoricals.  But this should resolve the issues with missing data handling.

File: statsmodels/regression/tests/test_lme.py [hunk 1]
```diff
@@ -726,6 +726,62 @@ def test_random_effects():
...
+                result1 = model1.fit()
+
+                # MixeLM handles missing
+                model2 = MixedLM.from_formula(
+                    fml, groups="g", data=df, missing='drop', **kwargs)
+                result2 = model2.fit()
+
+                assert_allclose(result1.params, result2.params)
+                assert_allclose(result1.bse, result2.bse)
+
```
Comment (josef-pkt) [inline statsmodels/regression/tests/test_lme.py]: maybe add some assert for `fittedvalues` or for some result arrays that have the nobs length

**Comment by josef-pkt** [comment]
Look fine to me, but I'm not familiar enough with mixedlm and missing value handling to spot possible problems. 

The renaming in the results is backwards incompatible, are they?  
names in summary are mostly irrelevant because users don't get direct access to them,  
but names in DataFrames would be used by users.

File: statsmodels/regression/mixed_linear_model.py [hunk 2]
```diff
@@ -853,19 +853,26 @@ def from_formula(cls, formula, data, re_formula=None, vc_formula=None,
...
             raise AttributeError("'groups' is a required keyword argument in MixedLM.from_formula")
+        groups = kwargs["groups"]
 
         # If `groups` is a variable name, retrieve the data for the
         # groups variable.
         group_name = "Group"
-        if type(kwargs["groups"]) == str:
-            group_name = kwargs["groups"]
-            kwargs["groups"] = np.asarray(data[kwargs["groups"]])
+        if type(groups) == str:
```
Comment (jbrockmendel) [inline statsmodels/regression/mixed_linear_model.py]: Does bytes/unicode matter here?

**Comment by kshedden** [comment]
Thanks, I think I have fixed this now.

**Pull request merged**